File: dummy.py
import time
import os
import signal
import argparse

NUM_LOOKUPS = 26

def signal_handler(signal_number, frame):
    pass

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Dummy accesses')
    parser.add_argument("--parent-pid", type=int)
    args = parser.parse_args()

    arr1 = list(range(NUM_LOOKUPS))

    signal.signal(signal.SIGUSR1, signal_handler)
    for i in range(NUM_LOOKUPS):
        os.kill(args.parent_pid, signal.SIGUSR1)
        signal.pause()

        n = arr1[i]

        os.kill(args.parent_pid, signal.SIGUSR1)
        signal.pause()
        print(f"Accessed {n} at arr1[{i}]")

File: dummy.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>

#define NUM_LOOKUPS 26

void signal_handler(int signum) {
    ;
}

int main(int argc, char *argv[]) {
    int nums[NUM_LOOKUPS];
    int ppid;
    
    printf("argv[1] = %s\n", argv[1]);
    sscanf(argv[2], "--parent-pid=%d", &ppid);
    printf("ppid = %d\n", ppid);

    signal(SIGUSR1, signal_handler); // Set signal handler for SIGUSR1
    for (int i = 0; i < NUM_LOOKUPS; i++) {
        kill(ppid, SIGUSR1);
        pause();

        int n = nums[i];

        kill(ppid, SIGUSR1);
        pause();
        printf("nums[%d] = %d\n", i, n);
    }
    return 0;
}